thriftbp/ttl_client.go

```diff
@@ -19,26 +25,50 @@ var _ Client = (*ttlClient)(nil)
 
 // ttlClient is a Client implementation wrapping thrift's TTransport with a TTL.
 type ttlClient struct {
-	thrift.TClient
+	// configs needed for refresh to work
+	generator ttlClientGenerator
+	ttl       time.Duration
+	jitter    float64
 
-	transport thrift.TTransport
+	replaceCounter metrics.Counter
 
-	// if expiration is zero, then the client will be kept open indefinetly.
-	expiration time.Time
+	// whole block guarded by lock
+	lock       sync.Mutex
+	client     thrift.TClient
+	transport  thrift.TTransport
+	expiration time.Time // if expiration is zero, then the client will be kept open indefinetly.
+	timer      *time.Timer
+	closed     bool
 }
 
 // Close implements Client interface.
 //
 // It calls underlying TTransport's Close function.
 func (c *ttlClient) Close() error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.closed = true
+	if c.timer != nil {
+		c.timer.Stop()
+	}
 	return c.transport.Close()
 }
 
+func (c *ttlClient) Call(ctx context.Context, method string, args, result thrift.TStruct) (thrift.ResponseMeta, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.client.Call(ctx, method, args, result)
+}
+
 // IsOpen implements Client interface.
 //
-// If TTL has passed, it closes the underlying TTransport and returns false.
-// Otherwise it just calls the underlying TTransport's IsOpen function.
+// It checks underlying TTransport's IsOpen first,
+// if that returns false, it returns false.
+// Otherwise it checks TTL,
+// returns false if TTL has passed and also close the underlying TTransport.
 func (c *ttlClient) IsOpen() bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
 	if !c.transport.IsOpen() {
```

Comment:
`return !c.closed && c.transport.IsOpen()`?

Reply:
also we can't return directly here, that doesn't handle the case that the connection is still good but the ttl was hit (`time.AfterFunc` could be off by a short period of time, or the case of background refresh failed but the original connection was still good).